compiler/semantic/scope.go

```diff
@@ -24,31 +30,72 @@ func (s *Scope) Exit() {
 	s.stack = s.stack[:len(s.stack)-1]
 }
 
-func (s *Scope) Bind(name string, ref dag.Op) {
-	s.tos().Define(name, ref)
+func (s *Scope) DefineVar(name string) (*dag.Var, error) {
+	b := s.tos()
+	if _, ok := b.symbols[name]; ok {
+		return nil, fmt.Errorf("symbol %q redefined", name)
+	}
+	ref := &dag.Var{
+		Kind: "Var",
+		Slot: s.nvars(),
+	}
+	b.Define(name, ref)
+	b.nvar++
+	return ref, nil
 }
 
-func (s *Scope) Lookup(name string) dag.Op {
+func (s *Scope) DefineConst(name string, def dag.Expr) error {
+	b := s.tos()
+	if _, ok := b.symbols[name]; ok {
+		return fmt.Errorf("symbol %q redefined", name)
+	}
+	val, err := kernel.EvalAtCompileTime(s.zctx, def)
+	if err != nil {
+		return err
+	}
+	if val.IsError() {
+		return fmt.Errorf("cannot resolve const %q at compile time", name)
+	}
+	literal := &dag.Literal{
+		Kind:  "Literal",
+		Value: zson.MustFormatValue(*val),
+	}
+	b.Define(name, literal)
+	return nil
+}
+
+func (s *Scope) Lookup(name string) dag.Expr {
 	for k := len(s.stack) - 1; k >= 0; k-- {
-		if e, ok := s.stack[k][name]; ok {
+		if e, ok := s.stack[k].symbols[name]; ok {
 			e.refcnt++
-			return e.op
+			return e.ref
 		}
 	}
 	return nil
 }
 
+func (s *Scope) nvars() int {
+	var n int
+	for k := len(s.stack) - 1; k >= 0; k-- {
```

Comment:
No need to iterate in reverse here.
```suggestion
	for k :=  range s.stack {
```